File: sandbox-go/exec_test.go
package sandbox

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestExecForwardsEnvPayload(t *testing.T) {
	t.Parallel()

	var seen map[string]any

	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/sandboxes/sandbox-1/exec" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if r.Method != http.MethodPost {
			t.Fatalf("unexpected method: %s", r.Method)
		}

		defer r.Body.Close()
		if err := json.NewDecoder(r.Body).Decode(&seen); err != nil {
			t.Fatalf("decode body: %v", err)
		}

		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"message":"Exec completed","data":{"stdout":"ok\n","stderr":"","exit_code":0,"duration_ms":6}}`))
	}))
	defer ts.Close()

	client, err := NewClient(ClientConfig{
		APIKey:  "test-key",
		BaseURL: ts.URL,
	})
	if err != nil {
		t.Fatalf("new client: %v", err)
	}

	_, err = client.Sandboxes.Use("sandbox-1").Exec(context.Background(), ExecInput{
		Cmd: "printenv HELLO",
		Env: map[string]string{"HELLO": "WORLD"},
	})
	if err != nil {
		t.Fatalf("exec: %v", err)
	}

	if seen["cmd"] != "printenv HELLO" {
		t.Fatalf("unexpected cmd: %#v", seen["cmd"])
	}

	env, ok := seen["env"].(map[string]any)
	if !ok {
		t.Fatalf("missing env payload: %#v", seen["env"])
	}
	if env["HELLO"] != "WORLD" {
		t.Fatalf("unexpected env payload: %#v", env)
	}
}

func TestRunCodeForwardsEnvPayload(t *testing.T) {
	t.Parallel()

	var seen map[string]any

	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/sandboxes/sandbox-1/code" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if r.Method != http.MethodPost {
			t.Fatalf("unexpected method: %s", r.Method)
		}

		defer r.Body.Close()
		if err := json.NewDecoder(r.Body).Decode(&seen); err != nil {
			t.Fatalf("decode body: %v", err)
		}

		w.Header().Set("content-type", "application/json")
		_, _ = w.Write([]byte(`{"message":"Exec completed","data":{"stdout":"WORLD\n","stderr":"","exit_code":0,"duration_ms":9}}`))
	}))
	defer ts.Close()

	client, err := NewClient(ClientConfig{
		APIKey:  "test-key",
		BaseURL: ts.URL,
	})
	if err != nil {
		t.Fatalf("new client: %v", err)
	}

	_, err = client.Sandboxes.Use("sandbox-1").RunCode(context.Background(), CodeInput{
		Language: CodeLanguagePython,
		Code:     `import os; print(os.getenv("HELLO"))`,
		Env:      map[string]string{"HELLO": "WORLD"},
	})
	if err != nil {
		t.Fatalf("run code: %v", err)
	}

	if seen["language"] != string(CodeLanguagePython) {
		t.Fatalf("unexpected language: %#v", seen["language"])
	}

	env, ok := seen["env"].(map[string]any)
	if !ok {
		t.Fatalf("missing env payload: %#v", seen["env"])
	}
	if env["HELLO"] != "WORLD" {
		t.Fatalf("unexpected env payload: %#v", env)
	}
}

File: sandbox/test/unit/exec.test.ts
import { describe, expect, test, vi } from 'vitest';

import { CodeLanguage } from '../../src/enums';
import { Sandbox } from '../../src/client';
import { ExecStream } from '../../src/streaming';

describe('Exec env support', () => {
  test('exec forwards env object', async () => {
    let requestBody: unknown = null;

    const fetchImpl = vi.fn(async (_url: URL | RequestInfo, init?: RequestInit) => {
      if (typeof init?.body === 'string') {
        requestBody = JSON.parse(init.body) as unknown;
      }

      return new Response(
        JSON.stringify({
          message: 'ok',
          data: {
            stdout: 'ok\n',
            stderr: '',
            exit_code: 0,
            duration_ms: 4,
          },
        }),
        {
          status: 200,
          headers: { 'content-type': 'application/json' },
        },
      );
    }) as typeof fetch;

    const client = new Sandbox({
      apiKey: 'test-key',
      fetchImpl,
    });

    await client.sandboxes.use('sandbox-1').exec({
      cmd: 'printenv HELLO',
      env: { HELLO: 'WORLD' },
    });

    expect(requestBody).toMatchObject({
      cmd: 'printenv HELLO',
      env: { HELLO: 'WORLD' },
    });
  });

  test('runCode forwards env object', async () => {
    let requestBody: unknown = null;

    const fetchImpl = vi.fn(async (_url: URL | RequestInfo, init?: RequestInit) => {
      if (typeof init?.body === 'string') {
        requestBody = JSON.parse(init.body) as unknown;
      }

      return new Response(
        JSON.stringify({
          message: 'ok',
          data: {
            stdout: 'WORLD\n',
            stderr: '',
            exit_code: 0,
            duration_ms: 8,
          },
        }),
        {
          status: 200,
          headers: { 'content-type': 'application/json' },
        },
      );
    }) as typeof fetch;

    const client = new Sandbox({
      apiKey: 'test-key',
      fetchImpl,
    });

    await client.sandboxes.use('sandbox-1').runCode({
      language: CodeLanguage.Python,
      code: 'import os; print(os.getenv("HELLO"))',
      env: { HELLO: 'WORLD' },
    });

    expect(requestBody).toMatchObject({
      language: CodeLanguage.Python,
      env: { HELLO: 'WORLD' },
    });
  });

  test('stream: true returns an ExecStream and strips client callbacks from request body', async () => {
    let requestBody: unknown = null;

    const fetchImpl = vi.fn(async (_url: URL | RequestInfo, init?: RequestInit) => {
      if (typeof init?.body === 'string') {
        requestBody = JSON.parse(init.body) as unknown;
      }

      const encoder = new TextEncoder();
      const stream = new ReadableStream<Uint8Array>({
        start(controller) {
          controller.enqueue(
            encoder.encode(
              'data: {"type":"stdout","data":"live\\n"}\n\ndata: {"type":"done","exit_code":0,"duration_ms":2}\n\n',
            ),
          );
          controller.close();
        },
      });

      return new Response(stream, {
        status: 200,
        headers: { 'content-type': 'text/event-stream' },
      });
    }) as typeof fetch;

    const client = new Sandbox({ apiKey: 'test-key', fetchImpl });
    const output = await client.sandboxes.use('sandbox-1').exec({
      cmd: 'echo live',
      stream: true,
      onStdout: () => undefined,
    });

    expect(output).toBeInstanceOf(ExecStream);
    expect(requestBody).toEqual({ cmd: 'echo live', stream: true });

    const logs = [];
    for await (const log of output) {
      logs.push(log);
    }

    expect(logs).toEqual([{ stream: 'stdout', data: 'live\n' }]);
  });

  test('onStdout streams live output and still returns ExecResult', async () => {
    const fetchImpl = vi.fn(async () => {
      const encoder = new TextEncoder();
      const stream = new ReadableStream<Uint8Array>({
        start(controller) {
          controller.enqueue(
            encoder.encode(
              'data: {"type":"stdout","data":"chunk\\n"}\n\ndata: {"type":"done","exit_code":0,"duration_ms":2}\n\n',
            ),
          );
          controller.close();
        },
      });

      return new Response(stream, {
        status: 200,
        headers: { 'content-type': 'text/event-stream' },
      });
    }) as typeof fetch;

    const client = new Sandbox({ apiKey: 'test-key', fetchImpl });
    const chunks: string[] = [];

    const result = await client.sandboxes.use('sandbox-1').exec({
      cmd: 'echo chunk',
      onStdout: (chunk) => {
        chunks.push(chunk);
      },
    });

    expect(chunks).toEqual(['chunk\n']);
    expect(result.stdout).toBe('chunk\n');
    expect(result.exit_code).toBe(0);
  });
});
